modal/functions.py

```diff
@@ -206,21 +207,19 @@ async def poll_function(self, timeout: Optional[float] = None):
 
     async def run_generator(self):
         data_stream = _stream_function_call_data(self.client, self.function_call_id, variant="data_out")
-        combined_stream = stream.merge(data_stream, stream.call(self.run_function))  # type: ignore
 
         items_received = 0
         items_total: Union[int, None] = None  # populated when self.run_function() completes
-        async with combined_stream.stream() as streamer:
-            async for item in streamer:
-                if isinstance(item, api_pb2.GeneratorDone):
-                    items_total = item.items_total
-                else:
-                    yield item
-                    items_received += 1
-                # The comparison avoids infinite loops if a non-deterministic generator is retried
-                # and produces less data in the second run than what was already sent.
-                if items_total is not None and items_received >= items_total:
-                    break
+        async for item in async_merge(data_stream, awaitable_to_aiter(self.run_function())):
```

Comment:
This should probably have an `aclosing` around the async_merge to be safe, since async_merge is an async generator with async finalizers.

`data_stream` also needs aclosing